File: CSharp/OpenTraceability/Utility/ObjectExtensions.cs
using Force.Crc32;
using System.Text;

namespace OpenTraceability.Utility
{
    public static class ObjectExtensions
    {
        public static Int32 GetInt32HashCode(this string strText)
        {
            Int32 hashCode = 0;
            if (!string.IsNullOrEmpty(strText))
            {
                hashCode = (int)Crc32Algorithm.Compute(Encoding.UTF8.GetBytes(strText));
            }
            return (hashCode);
        }

        public static Int16 GetInt16HashCode(this string strText)
        {
            Int64 i64Hash = strText.GetInt64HashCode();
            Int16 hash = (Int16)(i64Hash & 0xFFFF);
            hash ^= (Int16)((i64Hash >> 16) & 0xFFFF);
            hash ^= (Int16)((i64Hash >> 32) & 0xFFFF);
            hash ^= (Int16)((i64Hash >> 48) & 0xFFFF);
            return hash;
        }

        public static Int64 GetInt64HashCode(this string strText)
        {
            Int64 hashCode = 0;
            if (!string.IsNullOrEmpty(strText))
            {
                //Unicode Encode Covering all characterset
                byte[] byteContents = Encoding.Unicode.GetBytes(strText);
#pragma warning disable SYSLIB0021 // Type or member is obsolete
                System.Security.Cryptography.SHA256 hash = new System.Security.Cryptography.SHA256CryptoServiceProvider();
#pragma warning restore SYSLIB0021 // Type or member is obsolete
                byte[] hashText = hash.ComputeHash(byteContents);
                //32Byte hashText separate
                //hashCodeStart = 0~7  8Byte
                //hashCodeMedium = 8~23  8Byte
                //hashCodeEnd = 24~31  8Byte
                //and Fold
                Int64 hashCodeStart = BitConverter.ToInt64(hashText, 0);
                Int64 hashCodeMedium = BitConverter.ToInt64(hashText, 8);
                Int64 hashCodeEnd = BitConverter.ToInt64(hashText, 24);
                hashCode = hashCodeStart ^ hashCodeMedium ^ hashCodeEnd;
            }
            return (hashCode);
        }
    }
}

File: Java/OpenTraceability/src/main/opentraceability/utility/objectextensions.kt
package utility

import java.nio.charset.StandardCharsets
import java.security.MessageDigest

object ObjectExtensions {
    fun String.getInt32HashCode(): Int {
        val hashCode: Int
        if (isNullOrEmpty()) {
            hashCode = 0
        } else {
            val bytes = this.toByteArray(StandardCharsets.UTF_8)
            val digest = MessageDigest.getInstance("CRC32")
            hashCode = digest.digest(bytes).toInt()
        }
        return hashCode
    }

    fun String.getInt16HashCode(): Short {
        val i64Hash = this.getInt64HashCode()
        var hash: Short = (i64Hash and 0xFFFF).toShort()
        hash = hash.xor((i64Hash shr 16 and 0xFFFF).toShort())
        hash = hash.xor((i64Hash shr 32 and 0xFFFF).toShort())
        hash = hash.xor((i64Hash shr 48 and 0xFFFF).toShort())
        return hash
    }

    fun String.getInt64HashCode(): Long {
        val hashCode: Long
        if (isNullOrEmpty()) {
            hashCode = 0L
        } else {
            val bytes = this.toByteArray(StandardCharsets.UTF_16LE)
            val digest = MessageDigest.getInstance("SHA-256")
            val hashText = digest.digest(bytes)
            val hashCodeStart = bytesToLong(hashText, 0)
            val hashCodeMedium = bytesToLong(hashText, 8)
            val hashCodeEnd = bytesToLong(hashText, 24)
            hashCode = hashCodeStart xor hashCodeMedium xor hashCodeEnd
        }
        return hashCode
    }

    private fun bytesToLong(bytes: ByteArray, offset: Int): Long {
        var result: Long = 0
        for (i in 0 until 8) {
            result = result shl 8
            result = result or (bytes[offset + i].toLong() and 0xFF)
        }
        return result
    }
}
